k256/src/arithmetic/affine_op.rs

```diff
@@ -0,0 +1,423 @@
+use crate::arithmetic::mul::{G1, G2, MINUS_B1, MINUS_B2, MINUS_LAMBDA, Radix16Decomposition};
+use crate::arithmetic::projective::ENDOMORPHISM_BETA;
+use crate::arithmetic::scalar::{Scalar, WideScalar};
+use crate::{AffinePoint, FieldElement};
+use core::ops::{Add, Mul, Neg};
+use elliptic_curve::scalar::IsHigh;
+use elliptic_curve::subtle::{Choice, ConditionallySelectable};
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
+/// A wrapper around `AffinePoint` that provides additional ecc operations on affine points.
+pub struct PowdrAffinePoint(pub AffinePoint);
+
+impl Add<PowdrAffinePoint> for PowdrAffinePoint {
+    type Output = PowdrAffinePoint;
+
+    fn add(self, other: PowdrAffinePoint) -> PowdrAffinePoint {
+        if self.0.infinity != 0 {
+            return other;
+        }
+        if other.0.infinity != 0 {
+            return self;
+        }
+
+        if other.0.x == self.0.x {
+            if self.0.y == other.0.y {
+                return self.double();
+            } else {
+                //  x1 == x2 but y1 != y2 → vertical line → point at infinity
+                return PowdrAffinePoint(AffinePoint::IDENTITY);
+            }
+        }
+
+        // normalization is needed here to ensure the initial magnitude is 1.
+        let x1 = self.0.x.normalize_weak();
+        let y1 = self.0.y.normalize_weak();
+
+        let dx = other.0.x - x1;
+        let invert = dx.invert().unwrap();
+
+        let dy = other.0.y - y1;
+        let lambda = dy * invert;
+
+        let x3 = lambda.square() - x1 - other.0.x;
+        let y3 = lambda * (x1 + x3.negate(5)) - y1;
+
+        PowdrAffinePoint(AffinePoint {
+            x: x3.normalize_weak(),
+            y: y3.normalize_weak(),
+            infinity: 0,
+        })
+    }
+}
+
+impl Neg for PowdrAffinePoint {
+    type Output = PowdrAffinePoint;
+
+    fn neg(self) -> PowdrAffinePoint {
+        PowdrAffinePoint::neg(&self)
+    }
+}
+
+impl Mul<Scalar> for PowdrAffinePoint {
+    type Output = PowdrAffinePoint;
+
+    fn mul(self, other: Scalar) -> PowdrAffinePoint {
+        mul(&self, &other)
+    }
+}
+
+impl PowdrAffinePoint {
+    /// Ruduces the coordinates of the point to their canonical form.
+    pub fn normalize_coordinates(&self) -> Self {
+        PowdrAffinePoint(AffinePoint {
+            x: self.0.x.normalize(),
+            y: self.0.y.normalize(),
+            infinity: self.0.infinity,
+        })
+    }
+
+    /// Double the point.
+    pub fn double(self) -> PowdrAffinePoint {
+        if self.0.y.normalizes_to_zero().into() {
+            return PowdrAffinePoint(AffinePoint::IDENTITY);
+        }
+
+        let num = FieldElement::from(3u64) * self.0.x.square();
+        let denom = FieldElement::from(2u64) * self.0.y;
+        let lambda = num * denom.invert().unwrap();
+
+        let x3 = lambda.square() - FieldElement::from(2u64) * self.0.x;
+        let y3 = lambda * (self.0.x + x3.negate(3)) - self.0.y;
```

Comment:
```suggestion
        let num = self.0.x.square().mul_single(3);
        let denom = self.0.y.mul_single(2);
        let lambda = num * denom.invert().unwrap();

        let x3 = lambda.square() + self.0.x.mul_single(2).negate(2);
        let y3 = lambda * (self.0.x + x3.negate(4)) - self.0.y;
```

I believe this is cheaper: multiplying by small constants can be done without normalization (it does multiply the magnitude by that constant).

Reply:
Nice!